sktime/forecasting/compose/_networkpipeline.py

```diff
@@ -0,0 +1,278 @@
+                returned_arguments_kwarg[argument_name] = self._y
+                continue
+            if argument_value == "original_fh":
+                returned_arguments_kwarg[argument_name] = self._fh
+                continue
+
+            if type(argument_value) == list:
+                out = []
+                for list_val in argument_value:
+                    returned_step_value = self._check_steps_for_values(list_val)
+                    if returned_step_value is not None:
+                        out.append(returned_step_value)
+                returned_arguments_kwarg[argument_name] = out
+                continue
+            # go through all steps and look for returned values
+            returned_step_value = self._check_steps_for_values(argument_value)
+            if returned_step_value is not None:
+                returned_arguments_kwarg[argument_name] = returned_step_value
+
+        return returned_arguments_kwarg
+
+    def fit(self, y, X=None, fh=None):
+        self._set_y_X(y, X)
+        self._set_fh(fh)
+
+        for name, est, arguments in self._iter():
+            processed_arguments = {}
+
+            if "fit" in arguments and arguments["fit"] is None:
+                # this step must be skipped
+                continue
+            elif "fit" in arguments:
+                processed_arguments = self._process_arguments(arguments["fit"])
+            else:
+                processed_arguments = self._process_arguments(arguments)
+
+            if "y" in processed_arguments:
+                processed_arguments["y"] = check_y(processed_arguments["y"])
+            if "Z" in processed_arguments:
+                processed_arguments["Z"] = check_y(processed_arguments["Z"])
+            if "X" in processed_arguments and processed_arguments["X"] is not None:
+                processed_arguments["X"] = check_X(processed_arguments["X"])
+            if "X" in processed_arguments and processed_arguments["X"] is None:
+                del processed_arguments["X"]
+            # Transformers are instances of BaseTransformer and BaseEstimator
+            # Estimators are only instances of BaseEstimator
+
+            if hasattr(est, "fit_transform"):
+                t = clone(est)
+                out = t.fit_transform(**processed_arguments)
+                self._step_results[name] = out
+                self._fitted_estimators[name] = t
+            processed_arguments["fh"] = self._fh
+            # estimators have fit and predict methods
+            if hasattr(est, "fit") and hasattr(est, "predict"):
+                f = clone(est)
+
+                f.fit(**processed_arguments)
+                self._fitted_estimators[name] = f
+
+        self._is_fitted = True
+
+        return self
+
+    def _predict(self, fh=None, X=None, return_pred_int=False, alpha=DEFAULT_ALPHA):
+        if X is not None:
+            self._X = X
+        for name, est, arguments in self._iter():
```

Comment:
Do we iterate in reverse order here?

Reply:
No, `self._iter()` goes through the steps in the order in which they were created.

Reply:
Yes, you are right, `predict()` will need to be in reverse order. I missed this.

Reply:
Actually, thinking a little bit more about this, does predict need to always go through the steps in reverse order?

I am planning to ultimately use this pipeline for online one step ahead supervised classification. For this use case at the model fitting stage `y` is transformed, features are extracted, etc. The same transformations and feature extractions need to happen at the prediction stage as well in the same order as for `fit`.

For forecasting, let's assume that you fit your estimator after you transformed `y`. In this case, at the prediction stage would you not need to:

1. transform `y`, 
2. predict, 
3. inverse transform y.

I checked the code of  `TransformedTargetForecaster`, indeed the steps are executed in reverse order there. It will be quite easy to implement the same logic for the Network Pipeline but I just want to understand what we are trying to achieve. Is the pipeline supposed to be executing `predict()` if the estimator is a forecaster and `inverse_transform()` if it is a transformer? How do we handle the case if the forecaster was fitted on transformed `y`?

Reply:
I'm not sure about online one step ahead supervised classification (is that directional forecasting?). For forecasting, if the forecaster is fitted on transformed data, we first generated the predictions and then inverse transform the predictions. 

I think it's really important to write down what the simplest use case is that is enabled by this PR which is not possible with the `TransformedTargetForecaster` or a simple `PipelineForecaster`.

Reply:
yes, this is directional forecasting. 

The pipeline allows to execute any number of steps in almost any given order, as opposed to sequentially. This design was initially inspired by the directional forecasting use case in which you might need to extract different features from the original raw time series and labelling the observations based on some rule. For directional forecasts, I don't think this can easily be achieved with a simple pipeline. 

Regarding forecasting, maybe a simple use case of this pipeline would be to build some sort of an ensemble forecasting algorithm. For example, fitting two forecasters on the original `y` and having some sort of weighting or voting rule to produce an ensemble forecast. I can think a little bit more about it and propose some classification and forecasting examples for the documentation.